### PR TITLE
Allow Yarn workspace dependencies by default (and many other fixes)

### PR DESCRIPTION
**Pull request by chmeliik**
# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] New code has type annotations
- [ ] ~OpenAPI schema is updated (if applicable)~
- [ ] ~DB schema change has corresponding DB migration (if applicable)~
- [ ] ~README updated (if worker configuration changed, or if applicable)~
- [ ] Draft release notes are updated before merging

### CODE REVIEW

**Comment by chmeliik** [comment]
This fixes one of the many problems that yarn workspaces introduce. https://github.com/cachito-testing/cachito-yarn-workspaces-happy-path explains some of them.

Now, users *can*  at least get correct results by adding every workspace as a `file:` dependency to package.json

**Comment by chmeliik** [comment]
PTAL

This is the first part of Yarn workspace fixes - it allows `file:` dependencies in yarn.lock if they are workspaces.

The second part (**not part of this PR**) will be detecting workspaces even if they don't appear in yarn.lock

This PR also includes two related fixes:

1. Report `file:` dependencies for npm regardless of whether they are workspaces or not (previosuly, npm reported allowlisted dependencies but not workspaces)
2. Make the yarn allowlist behave the same as the npm allowlist (previously, yarn dependencies were allowlisted based on their path, not their name like npm)

**Comment by brunoapimentel** [comment]
> This is the first part of Yarn workspace fixes - it allows `file:` dependencies in yarn.lock if they are workspaces.

Just curious, but does this scenario happen at all? I'm asking because the sample repo we got from this ticket had no mention of workspaces in the yarn.lock file.

**Comment by brunoapimentel** [comment]
> > This is the first part of Yarn workspace fixes - it allows `file:` dependencies in yarn.lock if they are workspaces.
> 
> Just curious, but does this scenario happen at all? I'm asking because the sample repo we got from this ticket had no mention of workspaces in the yarn.lock file.

Nevermind, I just realized there are several instances of `file:` in the sample repository. I remember now the original state of the `package.json` when the user reached us: https://github.com/xperimental/kibana/commit/edcfd82cbdd2d00c65745d6aad3a9c80ea43cf1b

When you declare a dependency that is a workspace, but only use its version instead of `file:path`, this dep does not show up in the lockfile. Just a heads up that this is apparently a valid use case, so we will need to treat it in the following PR.

**Comment by chmeliik** [comment]
> When you declare a dependency that is a workspace, but only use its version instead of `file:path`, this dep does not show up in the lockfile. Just a heads up that this is apparently a valid use case, so we will need to treat it in the following PR.

Yup,  I have an integration test ready for that in https://github.com/cachito-testing/cachito-yarn-workspaces

Just trying to make the code less ugly now :smile:

File: tests/test_workers/test_pkg_managers/test_general_js.py [hunk 1]
```diff
@@ -794,6 +801,88 @@ def test_process_non_registry_dependency_github_not_in_nexus(mock_unrd, mock_gnc
...
+            "is_allowed": False,
+        },
+        {
+            "name": "not_a_file_dependency",
+            "version": "git+https://...",
+            "is_allowed": True,
+        },
+    ],
+)
+def test_vet_file_dependency_allowlist(test_input: dict[str, Any]):
```
Comment (brunoapimentel) [inline tests/test_workers/test_pkg_managers/test_general_js.py:872]: Do we need to annotate the return value here?

File: tests/test_workers/test_pkg_managers/test_yarn.py [hunk 2]
```diff
@@ -422,18 +480,16 @@ def mock_nexus_replacement_getitem(key):
     )
 
 
-@mock.patch("cachito.workers.pkg_managers.yarn._convert_to_nexus_hosted")
-def test_get_deps_disallowed_file_dep(mock_convert_hosted):
+def test_get_deps_disallowed_file_dep():
```
Comment (brunoapimentel) [inline tests/test_workers/test_pkg_managers/test_yarn.py:483]: Do we need to annotate the return value here?

File: tests/test_workers/test_pkg_managers/test_general_js.py (hunk 1 above)
Comment (chmeliik) [reply tests/test_workers/test_pkg_managers/test_general_js.py]: Yup, good catch. And that's why we need stricter mypy /shakes_fist

**Review by brunoapimentel — APPROVED** [review]
LGTM :rocket: 
And thanks for all the fi~sh~xes

**Review by lkolacek — APPROVED** [review]
Looks amazing! Thank you Adam. Nice logic commits as always. :taco:

**Pull request merged**